### ISSUE TITLE
incorrect FMRC regridding of gappy starts (non-contiguous F initialization times)

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1663]
**Reported by @AndrewWittenberg on 23 Mar 2016 19:00 UTC**
See also 2381.

I created an FMRC aggregation using the attached ens.agg file, then ran the following commands:

```
 	NOAA/PMEL TMAP
 	FERRET v6.97 (beta/opt)
 	Linux 2.6.32-573.18.1.el6.x86_64 64-bit - 03/09/16
 	23-Mar-16 14:23     

yes? use ens.agg
   dmgetting files...
building aggregation for member 01...
building aggregation for member 02...
aggregating 2 ensemble members...
yes? can v; set v ul; shade/y=0/x=140w/m=1 t_surf
yes? set v ll; shade/y=0/x=140w/m=1/lev t_surf[gt(tf_times)=tf_cal_t@fmrc]
yes? set v lr; shade/y=0/x=140w/m=1/lev t_surf[gt(tf_times)=tf_cal_t,gf(tf_times)=tf_lag_f]
```

Now with the following uncommented in ens.agg, so that we start a forecast every month, the plots look fine (fcst_noskip.gif):

```
let init_months = "{01,02,03,04,05,06,07,08,09,10,11,12}"
```

But if we instead do

```
let init_months = "{01,04,07,10}"
```

so that we only read the 1-January, 1-April, 1-July, and 1-October starts, then we get fcst_noskip.gif.

In this latter case, the bottom-right panel ("forecast skill" view) is not right.  Some of the regridding may be getting done along the wrong axis.

What this bottom-right panel *should* look like is the following.  Its first column (lead 0.5 months) should traverse along the bottom-most diagonal of the "diagonal" view (panel at lower left) -- i.e.

dark green, light orange, reddish orange, etc.

rather than

dark green, light green, light orange, light orange, etc.

As it stands, it looks as if this first column of the "skill" view is pulling the first 3 times from the bottom of the first column of the diagonal view; the next three times from the 2nd column of the diagonal view; etc.  Not right.

Instead, each column of the "skill" view should have, traversing upward from the bottom of the column:

DATA, gap, gap, DATA, gap, gap, ...

i.e.

column 1: dark green, gap, gap, light orange, gap, gap, reddish orange, ...

column 2: light green, gap, gap, reddish orange, gap, gap, reddish orange, ...

And the columns of the "skill" view should be 1 month (~30 days) apart, not 90 days apart, along the horizontal axis.

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/2391

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 28 Mar 2016 20:04 UTC**
In configuring a gappy FMRC, Ferret deduces the T-delta between the forecasts by finding the smallest T-delta between the consecutive files.  If this gap is not an integer divisor of all other gaps, it adjusts it by fractions (allows only 1/2 then 1/3) to find the true implied gap.  This logic had a bug (in SORT_FMRC_DSETS) that has now been fixed.  It was getting the wrong delta.

Even with this logic repaired, it remains possible to get the "wrong" conclusion about the natural T-delta between forecasts  e.g.  for "yes? FMRC = 1, 3, 5, 7, 9"  Ferret will see this as having no gaps at all.  It appears simply to have a uniform double-sized T-delta between forecasts.  If this situation is a problem, then we will need to add a qualifier FMRC/deltaF=<value>.

A slightly mor obscure case of the preceding is
 "yes? FMRC = 1, 5, 7, 9"
where Ferret will see just a single file missing between member no. 1 and member no. 5.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 29 Mar 2016 21:50 UTC**
Rather than ```FMRC/deltaF=<value>```, I'd suggest the proposal in #2393.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 15 Apr 2016 17:31 UTC**
In v6.97beta5 dated 4/14/16, the compact and diagonal views look correct now.

But for the gappy starts, the skill view is still not right; the data are being traversed incorrectly.

See the attached plots.  plot_month.gif looks perfect -- its skill plot (bottom right) pulls its leftmost column from the bottom diagonal of the diagonal plot (bottom-left), and its rightmost column from the top diagonal of the diagonal plot.  As we move up a given column in the skill plot, we move upwards & to the right along its corresponding diagonal in the diagonal plot.

Now plot_season.gif looks perfect for the left 2 panels.  But the skill plot (bottom right) is all wrong.  It should have 12 columns (not 4), corresponding to the 12 diagonals (12 leads) in the diagonal plot.  And each column should have only 8 elements (8 verification times) covering the span of 24 months in the vertical.  In the skill plot the horizontal cell-centers should be a month apart (not 3 months apart), and in the vertical the cells should be a month high (as they currently are).

Traversing upwards along a column in the skill plot, we should have: valid month, gap month, gap month, valid month, gap month, gap month, valid_month, etc.  Just imagine taking the bottom diagonal (i.e. the lead time of 0.5 months) of the diagonal plot, and sliding all its cells (whether filled or empty) to the left side; that'd be the first column of the skill plot.

Now repeat for the 2nd-from-bottom diagonal of the diagonal plot.  All the valid cells and gaps are shifted upwards by one cell, relative to the first diagonal.  So that'll also be the case for the 2nd column of the skill plot.

So what we should see in the skill plot, is a stack of 8 diagonal stripes flowing upwards to the right, separated by 2 gap-months in the vertical and 2 gap-months in the horizontal.  The sequence of colors from bottom-left to top-right along each diagonal stripe in the skill plot, will be identical to the sequence of colors traversing upwards along a column in the compact plot or diagonal plot.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 21 Apr 2016 00:29 UTC**
I have a theory about what is going on here, and I want to run it past you before looking further.

Back story:

    If you define two (or more) different FMRC's in the same session, the names of the special regrid axes differ, because they are distinct axes.  The first will get axes TF_CAL_T, TF_LAG_F and the next will get TF_CAL_T1, TF_LAG_F1, etc.  To see the special axis associated with each FMRC use SHOW DATA/ATTR.

    In this sample session I have created both a monthly and a seasonal FMRC using some of your files and we see:

```
        yes? show data/att
             currently SET data sets:
           37> andrew_fmrc     Forecast aggregation
         Variable    VarType(in dset)    AttributeName   AttType     Size OutFlag AttValue
        ------------------------------------------------------------------------------------------
         
         .                               aggregate name  CHAR        11   F       andrew_fmrc
                                         _cal_t_ax       CHAR        8    T       TF_CAL_T
                                         _lag_f_ax       CHAR        8    T       TF_LAG_F
         
         TAU_X                 FLOAT     long_name       CHAR        17   T       zonal wind stress
        ...
                                         _ftime2d        CHAR        8    T       TF_TIMES
                                         _cal_t_ax       CHAR        8    T       TF_CAL_T
                                         _lag_f_ax       CHAR        8    T       TF_LAG_F
        ...

        ...
           38> fmrc_seasonal  (default)  Forecast aggregation
         Variable    VarType(in dset)    AttributeName   AttType     Size OutFlag AttValue
        ------------------------------------------------------------------------------------------
         
         .                               aggregate name  CHAR        13   F       fmrc_seasonal
                                         _cal_t_ax       CHAR        9    T       TF_CAL_T1
                                         _lag_f_ax       CHAR        9    T       TF_LAG_F1
         
         TAU_X                 FLOAT     long_name       CHAR        17   T       zonal wind stress
        ...
                                         _ftime2d        CHAR        8    T       TF_TIMES
                                         _cal_t_ax       CHAR        9    T       TF_CAL_T1
                                         _lag_f_ax       CHAR        9    T       TF_LAG_F1
```

Some comments relative to this:

    o we are currently using the "full control" syntax, which we manually tell Ferret how to to the regridding.  Recall the document we wrote last Spring, where we spelled out a more automated syntax using a "REMAP" directive.  The analogous automated behaviors (and syntax) for FRMCs are still needed.  e.g.  FILL/remap=skill taux    or whatever.  This will tell Ferret to use the special axis that is indicated in the attributes.

   o I think it will help our ability to communicate eficiently on these matters if we have some shared datasets/scripts.  I have attached the benchmark, bn_aggregate_f.jnl (and .sub), which creates and then uses test files.  (The variables in the test files are made needlessly complex ... a choice I have come to regret, but so it goes ...)  Feel free to hack on these whenever you want to communicate/discuss an issue.  Send me back modified versions (or other scripts) to facilitate communication -- or upload them to trac.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 21 Apr 2016 20:38 UTC**
I'm attaching a simplified, self-contained set of benchmark scripts, and plots illustrating the above issues.

plot_monthly.gif looks good.

plot_seasonal.gif looks good in the left two panels, but not the bottom right.

plot_monthly_masked.gif is a version of plot_monthly.gif showing only the 1-JAN, 1-APR, 1-JUL, and 1-OCT starts.  Its bottom-right panel is what we ought to see in plot_seasonal.gif.

Note that the left panels in plot_monthly_masked.gif differ from plot_seasonal.gif.  In plot_seasonal.gif, we haven't told Ferret anything about the desired cell widths along the initialization axis -- so it has to infer it from the spacing of the initializations.  This indeed is the desired behavior.

Scripts: do_all.jnl, which calls test_fmrc.jnl, which calls make_dsets.jnl.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 22 Apr 2016 00:21 UTC**
Thank you so much for sending such a clear example.  I do, indeed, see where you are coming from.

BUT -- we will need to talk this over next week and think it through together.  My opening position is this:  The way that Ferret is handling things now is a little different from what you are envisioning -- but is self-consistent, defensible, and easier to understand visually.  (Yes, I am prepared to be persuaded otherwise.  I hope you are, likewise.)  Here's why ..

Many forecast model runs routinely have delta-L>1 between forecast runs, do they not?  For example, (making this sh*t up!) an operational weather model may output hourly with a 72 hour forecast and be run again 4 times daily -- ie. advancing 6 time steps between forecast runs.  So whereas your "seasonal" use case poses our problem as one where 2 out of three of the monthly model runs are **missing**; most use cases would pose these forecasts run series as **complete representations** -- no voids.  Delta-L>1 is merely a normal property of that particular FMRC.

Your proposed SKILL visualization:
   In the visualization that you propose (ref plot_monthly_masked.gif) our skill plot will give us the most highly resolved interpretation of forecast lead time, but the plot will be '''mostly voids'''.  Using the weather forecast example, each 6 time steps on the TF_CAL_T (vertical) axis will correspond to a diagonal matrix of 6 points on the TF_LAG_F (horizontal) axis.  i.e. 5 out of 6 of all cells will be void.  It is difficult to visualize the environmental processes you are seeing this way (the plot_monthly_masked.gif SKILL plot illustrates how difficult).  A FILL plot is not even possible

The current SKILL visualization:
   The visualization that Ferret is currently generating takes that very same 6x6 sub-matrix, but it smears the colors from the diagonal cells from left to right so they fill a broader 6x1 sub-matrix.  This definitely helps you to visualize the scientific phenomena more easily.  But in addition to being easier to view, it is also scientifically defensible.  The reality is that there is only a single forecast model run every 6 hours.  So the forecast lead times are less resolved concept as a result.   A very knowledgeable user like yourself would know that the most accurate interpretation of the wide 6x1 matrix, is to imagine a line that lies along its diagonal. 

Here's my proposal:
   I will make an easy-to-use tool that generates the high-resolution, sparse SKILL view that you like from the current SKILL view.  Your do_all.jnl already shows one approach -- it isn't a hard problem (TDB together).  And we will document the why's, wherefore's and useage of this tool in the User guide.  This means that for the precision-oriented user we have a complete solution.  And for the plebs (in which category I include myself!) the default visualization will be pleasing to the eye, albeit with a fuzzier interpretation of the F-lag resolution.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 22 Apr 2016 18:10 UTC**
I understand your point about visualization, but I think there's a much better way to address it -- will describe at the end of this comment.

First as to correctness...  Yes, the plot_monthly_masked.gif is mostly voids -- but that's the only truly correct representation of the data.

In plot_monthly_masked.gif, in the skill plot (bottom right), let's look at the dark-red cell in the lower-left corner.  That's the forecasted monthly-mean temperature for the month spanning t=1-jan-1980:1-feb-1980, and for the month spanning f=0:1 "months since initialization at 1-jan-1980."

Now examine the next cell up, which is a void.  This is because there was NO FORECAST of the t=1-feb-1980:1-mar-1980 temperature, for F=0:1 months lead.  But there *is* a forecast of t=1-feb-1980:1-mar-1980 temperature at F=1:2 months lead; it's the red box just to the right of this void.

There's an important distinction between the "compact" & "diagonal" plots and the "skill" plot.  In the first two, the cells really "ought" to have 0 width along the abscissa, since they were initialized at a single point in time (not a time range).  However, since it'd be impossible to see an infinitesimally-narrow ribbon (and there's no way to create a "delta function" in Ferret for integrating/averaging), for convenience we assign these cells a finite width, stretching halfway to the next cell (unless we have additional information, e.g. a 1-feb forecast, that'd motivate a 1-month rather than 3-month cell width).  But in the "skill" plot, the cells really, correctly, *ought* to have identical finite widths in both the vertical & horizontal directions.  In the vertical we have a time average over a range of actual dates (e.g. 1-jan-1980:1-feb-1980); in the horizontal we have a time average over a range of lead times (e.g. 0:1 months since initialization at 1-jan-1980).

Now look at the skill plot as currently implemented in plot_seasonal.gif.  Here the bottom-left red cell makes no sense.  It says there's a forecast of the 1-jan-1980:1-feb-1980 mean, that spans 0:3 months since initialization -- clearly impossible.

So back to the correct gappy representation.  We needn't stop here, since Ferret can easily interpolate across these voids.  I've updated do_all.jnl and plot_monthly_masked.gif to show one simple method (@fln followed by @spz); the result is added to plot_monthly_masked.gif as a new top-right panel.  We start with something correct, then explicitly say how we're imputing the gaps.  As you can see, it does an excellent job of reconstructing the bottom-right panel of plot_monthly.gif.  (And later if we so desired, we could fill some of these gaps by running actual forecasts initialized from 1-feb-1980 and 1-mar-1980; then the interpolation would be able to do less work.)

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 25 Apr 2016 16:12 UTC**
Your interpolation SKILL figure looks excellent. The added F resolution makes it much clearer.  (While driving I had imagined this plot, which was what changed my mind and lead to my email from Gig Harbor.  Sorry I didn't catch you 30 minutes sooner ...)

You may or may not agree with me on the following, but I believe that the '''interpolated''' SKILL view will be by far the dominant usage of it.  The sparse view needs to exist primarily as a reminder to the user of which cells do -- or not -- deserve the full level of trust.  When calling for forecast SKILL views from, say, an LAS interface, we'd probably only ever show the interpolated view.

Which means I'd like to put the interpolation steps directly into the code -- not requiring extra LET definitions.  So what is a nice syntax to differentiate the sparse and filled views?  In the full control syntax we might consider

```
gf(TF_TIMES)=TF_LAG_F@FILL
   vs
gf(TF_TIMES)=TF_LAG_F@FMRC  ! "FMRC" is the default ... usually omitted
```

In the as-yet-not-implemented automated syntax it might be something like

```
var[  vs
var[SKILL_SPARSE](SKILL]
)
```

We can talk about this on Wed., Thu.  We don't need to decide on the exact syntax now.

Which raises the question of what the "best" form of interpolation would be?  Would you agree that we can make a pretty sound argument for a linearly interpolated fill along the first column? That interpolation is equivalent to assuming that our sequence of initialization states has captured the phenomena. Once that first column is in place, a linear interpolation fill along the F (horizontal) axis is simply filling in the "missing" forecasts for the same verification time. (The left-most sub-matrices, where the first column has been filled by interpolation, is a little more complicated to describe/justify.)

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 25 Apr 2016 17:37 UTC**
Yes, I agree that in such gappy cases, the user will usually proceed to the interpolated view -- and it'd be great to have a simple way to do this interpolation, per your proposed syntax.

I do feel that the gappy version should be the default.  If the user must explicitly request filling (even via simple syntax), then they're forced to know at least a little about what they're doing, and how to interpret the resulting plot.

It'd be nice to have the interpolation be bilinear, e.g. as done by CONTOUR and FILL but on the grid-scale.  As an aside, that'd be a great general-purpose transformation to add: 

```
shade gappy[equivalently

```
shade gappy [tf=@fbl](t=@fbl,f=@fbl]
```

or)
```

where ```@fbl``` stands for "fill bilinear".  This could reduce to @lin when performed on only a single axis, and perhaps generalized to multilinear interpolation when applied to 2 or more axes simultaneously.  Other multi-axis interpolation methods (e.g. bicubic, spline) can be found in Numerical Recipes:

[interpolation is a big enough topic that it probably deserves a new ticket.

Short of multi-axis interpolation, the easiest thing would be to just do ```[t=@fln](http://www.aip.de/groups/soe/local/numres/bookcpdf/c3-6.pdf]

Multi-axis)``` to fill each column.  Or fancier, do what I did in plot_monthly_masked.gif: fill in T, then smooth in F, then add back the cells lost at the edges:

```
let skill = temp[missing(skill[t=@fln,f=@spz:3](gt(tf_times)=tf_cal_t,gf(tf_times)=tf_lag_f]
shade),skill[t=@fln])
```

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 11 May 2016 16:32 UTC**
Here is an outline of the code changes that will be needed to generate a full-resolution, sparse lag-F axis, per the discussion above:

Full control syntax: "sst[   o testing scripts
        o ~hankin/FERRET32/fer/demo_fmrc.jnl
        o bn_aggregate_f.jnl
        o bn_aggregate_tef.jnl
    o create the lag-F axis (named "TF_LAG_F", "TF_LAG_F1", ...)
        o in fmt/src/cd_make_fmrc_axes.F see "NOW FINALLY THE FORECAST LEAD TIME AXIS OF THE FMRC"
        o in the call to TM_MAKE_BASIC_AXIS change the delta time to match the delta time on the FMRC lag-T axis
        o see note on "Monthly data in FMRCs" below
    o request the data needed to perform the regrid specified by the user
        o stk/aux_regrid_lims_2d_tf.F
        o quite possibly no change is required in the request for data
    o use the received data to perform the regrid
        o doo/do_aux_var_regrid_2d_tf.F
        o make sure to pre-fill with missing values, since the result will now be sparse
        o look for "N-lag loop"
            o I believe that the F-lag loop is referred to as the "N-lag loop" in this code.  I believe the name refers to "N actual forecasts ahead".  We are switching to an outlook in which we are thinking in terms of "N potential forecasts ahead" (with some of the potential forecasts missing), which is what leads to the sparse result
    o add logic for auto-filling (interpolation) of the sparse field
        o need to decide upon a syntax for this.  The "full control" syntax is "sst[gt(tf_times)=TF_CAL_T@FMRC,gf(tf_times)=TF_LAG_F@FMRC](gt(tf_times)=TF_CAL_T,gf(tf_times)=TF_LAG_F]"

)" where the "@FMRC" transformation type as the default if omitted.  My impulse would be that the auto-filled result should be indicated by a syntax such as "sst[gt(tf_times)=TF_CAL_T,gf(tf_times)=TF_LAG_F@FILL]"
        o note: the cx_regrid_trans value for that "@FMRC" algorithm is named with parameter "pauxrgrd_fmrc".  You can grep the code to find the places that are relevant -- creating automatic labels, etc.
        o My coding suggestion would be to create a new parameter pauxrgrd_fmrc_fill, and a new subroutine that performs the auto-fill/interpolation to be called from doo/do_aux_var_regrid_2d_tf.F if the @FILL transformation has been requested

'''Monthly data in FMRCs -- notes on the messiness of monthly data, in which there is no true delta-T
'''
    o In an ideal world it would be desirable to explicitly capture the ambiguous notion of a "regular" monthly time series in the FMRC code.  In other words, to support an underlying encoding of 1, 2, 3, ... referring to month number (from some origin).  The current code, instead, sticks with the CF time encoding (days, hours, whatever) and allows small errors in regularity of spacing.  Messy and imperfect.  Besides cleaner code, another nice outcome would be that plot labels would then be in units of months (e.g. showing a forecast ability 3 months in advance).
    o Ansley has new code (which I have not seen) that might provide a nice way to bring the concept of "regular" into the FMRC's.  It would be worth looking over how much work it would be to incorporate this into the FMRC framework and deciding if the effort is worthwhile.
        o presumably with Ansley's code the incoming time series axes of the forecasts will be encoded (somehow) as month numbers
        o you'd need to generate lag-T, cal-T, lag-F and cal-F all correspondingly in units of months (simple changes)
        o The tf_times array would need to be converted into units of months (and would then become beautifully simple and regular small integer month numbers).  The code that assembles the 2d time array, tf_times, already includes the logic complexity of allowing for the T0 time encoding origin to shift between model runs.  You would need only modify this logic.
        o in several places the numerical code that depends upon the 2d time array could then be simplified and cleaned up
        o TDS special case: Note that FMRCs may be created by TDS, as an alternative to being synthesized by Ferret.  Separate code would be needed to address monthly data in this case.  See ~hankin/FERRET32/fer/demo_fmrc.jnl for an example script that can be run on either a TDS-generated or a Ferret-generated FMRCs.  (I dont think there is a benchmark test of a TDS-generated FMRC.  With Caron gone from Unidata, I'm not sure what the future of TDS-generated FMRCs is.)

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 11 May 2016 16:34 UTC**
added karl as cc in case not already getting this

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 11 May 2016 16:35 UTC**
**really* added karl as cc in case not already getting this

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 23 Mar 2016 19:00 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/ens.agg

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 23 Mar 2016 19:01 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/fcst_noskip.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 23 Mar 2016 19:01 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/fcst_skip.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 15 Apr 2016 17:32 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/plot_month.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 15 Apr 2016 17:32 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/plot_season.gif

**Comment by karlmsmith** [comment]
**Attachment from steven.c.hankin on 21 Apr 2016 00:30 UTC**
F-aggregation benchmark script
REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/bn_aggregate_f.jnl

**Comment by karlmsmith** [comment]
**Attachment from steven.c.hankin on 21 Apr 2016 00:31 UTC**
'subroutine' script for F-aggregation benchmark
REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/bn_aggregate_f.sub

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 21 Apr 2016 20:39 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/plot_monthly.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 21 Apr 2016 20:39 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/plot_seasonal.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 21 Apr 2016 20:40 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/make_dsets.jnl

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 21 Apr 2016 20:40 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/test_fmrc.jnl

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 22 Apr 2016 18:11 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/do_all.jnl

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 22 Apr 2016 18:12 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/plot_monthly_masked.gif

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 25 Apr 2016 17:37 UTC**

REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_2391_GIT_1663/do_all.2.jnl

**Comment by karlmsmith** [comment]
Have this working (not checked in) for the case where the forecasted (verification) and initialization axis are regular and one of the initialization time matches one of the forecasted times for the first forecast.  I make a minimal superset of regular axis coordinates that contain times for both the forecasted and initialization times, and the TF_LAG_T, TF_CAL_F, TF_LAG_F, TF_CAL_T, and TF_TIMES all use these coordinates for their axes.  (So if a forecast is made every 3 t_unit and a forecasted times are every 2 t_unit, the superset coordinates will be every t_unit.)  The original aggregated data is still defined on its own grid, so the first step in using this data is to regrid this data to TF_LAG_T, TF_CAL_F
```
DEFINE DATA/AGGREGATE/F testdata = ....
! contains aggregated variable "actual"

LET actualfine = actual[GT=TF_LAG_T@BIN,GF=TF_CAL_F@BIN]
! plot shows something like plot_monthly_masked top left

LET diagview = actualfine[gt(tf_times)=tf_cal_t]
! plot shows something like plot_monthly_masked bottom left

LET leadview = actualfine[gt(tf_times)=tf_cal_t,gf(tf_times)=tf_lag_f]
! plot shows something like plot_monthly_mashed bottom right
```

The restriction of the regular axes and an initialization time matching a first forecast forecasted time is just to automatically come up with a minimal superset of coordinates, but also makes it easy to generated the TF_... values.

Question: is this restriction a big problem?  If not, I am thinking there will then need to be something (probably a new command) that the user provides the coordinates which then generates the TF_.... axes and TF_TIMES.

**Comment by karlmsmith** [comment]
Example plots using four forecasts of "monthly" (365.2425/12 days) made every three "months" from the benchmarks.

![forecast_actual](https://user-images.githubusercontent.com/16406727/41423645-b70528f2-6fb0-11e8-88a1-f2768a36def7.gif)
![forecast_actualfine](https://user-images.githubusercontent.com/16406727/41423663-c4415202-6fb0-11e8-84bc-8de0cedd941c.gif)
![forecast_diagview](https://user-images.githubusercontent.com/16406727/41423672-c9361bda-6fb0-11e8-9661-07dd380cfc6f.gif)
![forecast_leadview](https://user-images.githubusercontent.com/16406727/41423677-cd418cc8-6fb0-11e8-8d54-8f5a956d42e8.gif)

**Comment by karlmsmith** [comment]
FYI - the data for "actual" are perfect forecasts, so some of the test is making sure values/colors are identical where expected.

**Comment by karlmsmith** [comment]
Changes commited https://github.com/NOAA-PMEL/Ferret/commit/ea0d1ee7f53974b8b4408b98ba3c2b7c8e855aa0
which require regular forecasted time and forecast initialization axes that share some dates in order to auto-generated the superset axis coordinates which is used to create the TF_... axes and TF_TIMES variable.

As mentioned in ticket #1665, allowing the user to specify the desired coordinates for the TF_... axes and resulting TF_TIMES variable would be desirable and remove the above restrictions, although the above could remain the default.
